### ISSUE TITLE
Designing the policy for Atomic platforms

### ISSUE THREAD

**Issue by mgrepl** [mgrepl/seatomic#1]
This issue is created to have a discussion about a basic design of seatomic policy. Once we have some agreed basic concept here it will be a part of Wiki.

**Comment by mgrepl** [comment]
## seatomic policy

Initial draft.

### Process labeling
- One process type for all SYSTEM Processes (unconfined_domain) os_t or system_t?
- Docker process and container processes would use separate types. docker_t, spc_t and svirt_lxc_net_t; svirt_sandbox_file_t (Hopefully we can rename these to container_t, and container_file_t or container_content_t)

### File system labeling.
-  /usr, /bin, /sbin, /src would have a type that can be read/write/execute within the containers (usr_t)?
-  /\* would have types that are not available to the container at all IE default_t.
- /proc and /sys would have to have labels that containers can read. Probably will need to have some defined types.
- /dev would still be labeled as default_t, IE No access from the containers.
- /var/lib/docker would have labeling provided by the docker-selinux package.

**Comment by mgrepl** [comment]
https://bugzilla.redhat.com/show_bug.cgi?id=1265406

**Comment by rhatdan** [comment]
We would need to make sure that docker-selinux package would still work.  Currently we rely on virt.pp for this also.  Which we might want to also include, since  we might want to run svirt_t vms on an atomic host.

**Comment by ssekidde** [comment]
Will this also include kubernetes policy? 
https://github.com/eparis/kubernetes-selinux

**Comment by mgrepl** [comment]
I believe we would define what core components are going to have own (the current policy).

**Comment by rhatdan** [comment]
Well k8s will be shipped in a container in the future, so probably not.

**Comment by mgrepl** [comment]
Also another question is if the current flask is too complex or we need to stay with it as we have it.

**Comment by rhatdan** [comment]
Not sure what you mean.  If it was up to me I think we should attempt a rewrite in CIL.   Rethink everything.

**Comment by mgrepl** [comment]
I meant security classes and permissions.

**Comment by rhatdan** [comment]
I am fine keeping the same security classes and permissions.  But this is something we should open up to upstream SELinux and see if they have ideas of changes then made to SEAndroid that made it easier.  It would be nice if we end up with policy as simple as SELAndroid.

**Comment by wrabcak** [comment]
I like the idea to try reduce current flask for SEAtomic. I believe we could remove some classes and permissions. Definitely, we should discuss this with upstream.

**Comment by mgrepl** [comment]
The questions if it is applicable here. We still need to have 

systemd/journald/kernel/sshd/sssd/docker/virt

as a base and we still need to have a user login domain. I would like to set a base with a policies and ask about it on in upstream

**Comment by rhatdan** [comment]
I would prefer to not have systemd/journald/kernel/sshd/sssd as different domains, if possible.  system_t Then as use cases come up for additional domains, we could add them.  Not even sure if you would want a separate user domain unless it is just unconfined_t.

**Comment by wrabcak** [comment]
Dan, 
After my last discussion with Mirek,  the concept of merging systemd/journald/kernel/sshd/sssd is actual and will be implemented.

**Comment by cgwalters** [comment]
Is `system_t` an unconfined domain?

**Comment by wrabcak** [comment]
No.

**Comment by rhatdan** [comment]
Well it would be about as unconfined as you can get since it includes the kernel...  Whether or not we have the concept of unconfined might be questionable.

**Comment by cgwalters** [comment]
I'll just link to https://bugzilla.redhat.com/show_bug.cgi?id=1265406#c4

**Comment by mgrepl** [comment]
It is about system_t being unconfined domain (hopefully not) against a login user domain. The problem with system_t for docker,virt is with a compatibility.

For container types we will use CIL namespaces.

**Comment by rhatdan** [comment]
Right, I don't think docker and libvirt should be system_t, they should probably be in their own domains.  We have had several CVEs from docker just downloading images, it would be nice to run docker as a confined domain to at least mitigate these problems.  Preventing an image from attacking the host.

**Comment by stephensmalley** [comment]
I think Dan's last comment highlights why it would be a mistake to only have a single system_t domain. In the Android model, we tightly confine each system process (which protects them from malicious inputs, executing unauthorized code, making their memory executable, etc).  Confinement is not merely to prevent bad actors from doing bad things, but to protect good actors from being tricked into doing bad things. You can still get a small policy even with per-service domains because you only have a few services running directly on the host OS (i.e. outside a container) and everything within the container runs in a single domain.
As before, I'd encourage you to consider some of the Android policy security goals, as described in 
my LSS talks, e.g. http://kernsec.org/files/lss2014/lss2014_androidtcb_smalley.pdf and http://kernsec.org/files/lss2015/lss2015_selinuxinandroidlollipopandm_smalley.pdf and see whether they might be viable for Atomic, e.g. strictly enforcing read-only OS except for the authorized update process, enforcing W^X for files and memory as much as possible, etc.  The Android policy itself is quite small and easy to read as well (https://android.googlesource.com/platform/external/sepolicy), so you might have a look at that and think about whether you truly want to stay with refpolicy or start over.
I was also confused by the statement in the initial draft above that said that /usr, /bin, /sbin, and /src would be labeled with a type that is rwx by containers.  rx I can understand if they are shared, but not w.

**Comment by mgrepl** [comment]
I agree with you guys here. Probably misunderstanding with some my comments.

**Comment by rhatdan** [comment]
/usr, /bin/, /sbin should be rx not rwx.  We currently do this in atomic host from a DAC point of view.  /usr is actually mounted Read/Only.

**Comment by mgrepl** [comment]
Ok I read the links from @stephensmalley. It looks good and I will prepare similar presentation. 

We wanted to go with different ideas here and have really minimal number of types. But maybe it is not so critical because there is a minimal number of services on Atomic.

Also I believe we want to continue with CIL because we want to use namespaces for container policies.

But still we can start with kernel_t, init_t, os_t + services around containers as the base and then just add confinements.

**Comment by mgrepl** [comment]
Note. I am just trying to get a minimal CIL policy working on the Atomic host to see how it works with CIL.
